### ISSUE TITLE
[Ray: Air]  MaxAbsScaler to process data, and the processing speed is extremely slow

### ISSUE THREAD

**Issue by yiwei00000** [ray-project/ray#39150]
### What happened + What you expected to happen

Our data volume is 100000 pieces, each with 1000 columns, totaling approximately 1.5G. The data is stored on hdfs.
The Ray cluster has 6 cores, each with 24GB of memory, and a total of three nodes.

I use MaxAbsScaler's fit_transform It takes about 10 minutes for the transformer to process these data, but I only need 15 seconds to use Spark's MaxAbsScaler to fit and transform. Why is ray so slow? Spark and ray use the same machine resources.

### Versions / Dependencies

ray 2.6.0
hdfs 3.2.2

### Reproduction script

import ray
from pyarrow import fs

ray.init('auto')
hdfs_fs = fs.HadoopFileSystem.from_uri("hdfs://jd-hadoop/?user=root")
ds = ray.data.read_csv('/data/eps-files/', filesystem=hdfs_fs,parallelism=2000)
files = sorted(ds.input_files())
ds = ray.data.read_csv(files[0:100], filesystem=hdfs_fs,parallelism=100)
ds = ds.materialize()
ds.max("x1")

cols = ds.columns()[0:1001]
ds_1001 = ds.select_columns(cols)
ds_1001 = ds_1001.materialize()
from ray.data.preprocessors import MaxAbsScaler
preprocessor = MaxAbsScaler(columns=cols[1:len(cols)])
ds_1001_trans = preprocessor.fit_transform(ds_1001) # **It takes about 10 minutes**
ds_1001_trans = ds_1001_trans.materialize()
ds_1001_trans.show(10)

### Issue Severity

High: It blocks me from completing my task.

**Comment by scottjlee** [comment]
Hi @yiwei00000 , any reason to `materialize()` the dataset in the above code example? This will load the entire dataset in memory, as opposed to taking advantage of streaming execution.

**Comment by dragno99** [comment]
@yiwei00000 these preprocessors are not well written

this is the method from preprocessor class, you can see while using ds.map_batch, they are not passing num_cpus properly, you can check their code, thats why these things are using only single core by default

```
def _transform(
        self, ds: Union["Dataset", "DatasetPipeline"]
    ) -> Union["Dataset", "DatasetPipeline"]:
        # TODO(matt): Expose `batch_size` or similar configurability.
        # The default may be too small for some datasets and too large for others.
        transform_type = self._determine_transform_to_use()

        # Our user-facing batch format should only be pandas or NumPy, other
        # formats {arrow, simple} are internal.
        kwargs = self._get_transform_config()
        if transform_type == BatchFormat.PANDAS:
            return ds.map_batches(
                self._transform_pandas, batch_format=BatchFormat.PANDAS, **kwargs
            )
        elif transform_type == BatchFormat.NUMPY:
            return ds.map_batches(
                self._transform_numpy, batch_format=BatchFormat.NUMPY, **kwargs
            )
        else:
            raise ValueError(
                "Invalid transform type returned from _determine_transform_to_use; "
                f'"pandas" and "numpy" allowed, but got: {transform_type}'
            )

```



i was facing same issue with Categorizer, my current workaround is that, i just overrided few methods and it worked



```

class MyCategorizer(Preprocessor):

    def __init__(self, columns: List[str], dtypes: Optional[Dict[str, pd.CategoricalDtype]] = None, ):
        if not dtypes:
            dtypes = {}

        self.columns = columns
        self.dtypes = dtypes

    def _fit(self, dataset: Dataset) -> Preprocessor:
        columns_to_get = [column for column in self.columns if column not in self.dtypes]
        if columns_to_get:
            unique_indices = _get_unique_value_indices(dataset, columns_to_get, drop_na_values=True, key_format="{0}")
            unique_indices = {
                column: pd.CategoricalDtype(values_indices.keys())
                for column, values_indices in unique_indices.items()
            }
        else:
            unique_indices = {}
        unique_indices = {**self.dtypes, **unique_indices}
        self.stats_: Dict[str, pd.CategoricalDtype] = unique_indices
        return self

    def _transform(self, ds: Union["Dataset", "DatasetPipeline"]) -> Union["Dataset", "DatasetPipeline"]:
        # TODO(matt): Expose `batch_size` or similar configurability.
        # The default may be too small for some datasets and too large for others.
        transform_type = self._determine_transform_to_use()

        # Our user-facing batch format should only be pandas or NumPy, other
        # formats {arrow, simple} are internal.
        kwargs = self._get_transform_config()
        if transform_type == BatchFormat.PANDAS:
            return ds.map_batches(
                self._transform_pandas,
                batch_format=BatchFormat.PANDAS,
                batch_size=batch_size,
                num_cpus=4,
                # num_gpus=1,
                **kwargs
            )
        elif transform_type == BatchFormat.NUMPY:
            return ds.map_batches(
                self._transform_numpy,
                batch_format=BatchFormat.NUMPY,
                batch_size=batch_size,
                num_cpus=4,
                **kwargs
            )
        else:
            raise ValueError(
                "Invalid transform type returned from _determine_transform_to_use; "f'"pandas" and "numpy" allowed, but got: {transform_type}')

    def _transform_pandas(self, df: pd.DataFrame):
        df = df.astype(self.stats_)
        return df

    def __repr__(self):
        return (f"{self.__class__.__name__}(columns={self.columns!r}, "f"dtypes={self.dtypes!r})")
```

Now, i can use **MyCategorizer** instead of **Categorizer** class, you can also do something like this.

**Comment by raulchen** [comment]
Also the `Preprocessor` API is now deprecated, we suggest use `map_batches(preprocess_fn)` instead. 
I'm closing it for now, feel free to reopen if you still have issues with the new API and the latest Ray.

**Comment by Taurus-Le** [comment]
@raulchen Do you mean all of the preprocessors are deprecated? Am I supposed to implement all the transformations with map_batches myself?
